lib/charms/operator_libs_linux/v0/dnf.py

```diff
@@ -0,0 +1,207 @@
+            return None
+
+        full_version = [self.version, f"-{self.release}"]
+        if self.epoch:
+            full_version.insert(0, f"{self.epoch}:")
+
+        return "".join(full_version)
+
+
+def version() -> str:
+    """Get version of `dnf` executable."""
+    return _dnf("--version").splitlines()[0]
+
+
+def installed() -> bool:
+    """Determine if the `dnf` executable is available on PATH."""
+    return shutil.which("dnf") is not None
+
+
+def upgrade(*packages: Optional[str]) -> None:
+    """Upgrade one or more packages.
+
+    Args:
+        *packages (Optional[str]):
+            Packages to upgrade on system. If packages is omitted,
+            upgrade all packages on the system.
+    """
+    _dnf("upgrade", *packages)
+
+
+def install(*packages: Union[str, os.PathLike]) -> None:
+    """Install one or more packages.
+
+    Args:
+        *packages (Union[str, os.PathLike]): Packages to install on the system.
+    """
+    if not packages:
+        raise TypeError("No packages specified.")
+    _dnf("install", *packages)
+
+
+def remove(*packages: str) -> None:
+    """Remove one or more packages from the system.
+
+    Args:
+        *packages (str): Packages to remove from system.
+    """
+    if not packages:
+        raise TypeError("No packages specified.")
+    _dnf("remove", *packages)
+
+
+def fetch(package: str) -> PackageInfo:
+    """Fetch information about a package.
+
+    Args:
+        package (str): Package to get information about.
+
+    Returns:
+        PackageInfo: Information about package.
+
+    Notes:
+        `package` needs to exactly match the name of the package that you are fetching.
+        For example, if working with the `python2` package on select EL distributions,
+        `dnf.install("python2")` will succeed, but `dnf.fetch("python2")` will return
+        the package in ABSENT state. This is because the name of the python2 package is
+        python2.7, not python2. To get info about the python2 package, you need to use
+        its exact name: `dnf.fetch("python2.7")`.
+    """
+    try:
+        stdout = _dnf("list", "-q", package)
+        # Only take top two lines of output.
+        status, info = stdout.splitlines()[:2]
+
+        # Check if package is in states INSTALLED or AVAILABLE. If not, mark absent.
+        if "Installed" in status:
+            state = _PackageState.INSTALLED
+        elif "Available" in status:
+            state = _PackageState.AVAILABLE
+        else:
+            raise Error(f"{package} is not installed or available but {status} instead.")
+
+        pkg_name, pkg_version, pkg_repo = info.split()
+        name, arch = pkg_name.rsplit(".", 1)
+
+        # Version should be good, but if not mark absent since package
+        # is probably in a bad state then.
+        version_match = re.match(r"(?:(.*):)?(.*)-(.*)", pkg_version)
+        if not version_match:
+            raise Error(f"Bad version {pkg_version}. Marking {package} absent.")
```

Comment:
Shouldn't we just `return `PackageInfo(name=package)` here, instead of raising and re-catching the error? Also, might be better to be more explicit with the `state=ABSENT` field.

Reply:
Refactored in https://github.com/canonical/operator-libs-linux/pull/75/commits/f75e25fb6c7def8537366ebc0256396eee54d342. No longer recapturing error and ABSENT `_state` assignment is explicit rather than implicit.